Extensions/dnSpy.AsmEditor/MethodBody/CilBodyVM.cs

```diff
@@ -316,6 +317,32 @@ public LocalsIndexObservableCollection(CilBodyVM owner, Func<LocalVM> createNewI
 
 		bool ReplaceInstructionWithNopCanExecute(InstructionVM[] instrs) => instrs.Any(a => a.Code != Code.Nop);
 
+		void ReplaceInstructionWithMultipleNops(InstructionVM[] instrs) {
+			var old1 = InstructionsListVM.DisableAutoUpdateProps;
+			var old2 = DisableHasError();
+			try {
+				InstructionsListVM.DisableAutoUpdateProps = true;
+
+				// Speed optimization, see comment in SimplifyAllInstructions()
+				UninstallInstructionHandlers(instrs);
+
+				foreach (var instr in instrs) {
```

Comment:
Your code uses instr.Index which isn't valid after you add more instructions in the list.

To fix this you should sort `instrs` in reverse order (highest Index is 1st in the array) and then you can safely execute the rest of the code below. Before the loop: `Array.Sort(instrs, (a, b) => b.Index - a.Index);`

Reply:
Ah, good catch. I didn't notice anything wrong while testing so I thought it used IndexOf internally.